driver/src/orderbook/streamed/state.rs

```diff
@@ -206,9 +216,39 @@ impl State {
             .get_address_by_id(order.buy_token)
             .ok_or_else(|| anyhow!("unknown buy token"))?;
         let buy_balance = self.balances.entry((user_id, buy_token)).or_default();
-        buy_balance_fn(buy_balance)?;
+        buy_balance_fn(buy_balance)
+    }
 
-        Ok(())
+    fn apply_solution_submission(
+        &mut self,
+        event: &SolutionSubmission,
+        block_batch_id: BatchId,
+    ) -> Result<()> {
+        let fee_token = self
+            .tokens
+            .get_address_by_id(0)
+            .ok_or_else(|| anyhow!("solution without fee token"))?;
+        self.revert_last_solution(fee_token, block_batch_id);
+        self.last_solution.batch_id = block_batch_id;
+        self.last_solution.user_id = event.submitter;
+        self.last_solution.burnt_fees = event.burnt_fees;
+        self.balances
+            .entry((event.submitter, fee_token))
```

Comment:
Does this add a new entry to the balance "dict" if it doesn't already exist and do nothing otherwise?

Reply:
With the line below `or_default` yes. It is like "get the element or create a default if it doesn't exist and then get that".